### PR TITLE
Add warning when OR is found in JOIN

### PR DESCRIPTION
**Pull request by chi-tsai**
Test plan - 

Created and passed JUnit tests that check for warnings in simple SQL queries that involves having an OR in a JOIN. 
Created and passed JUnit tests that check no warnings appear when there is OR in JOIN in the SQL query, or if the OR is not directly in the predicate. 

```
== RELEASE NOTES ==

General Changes
* In StatementAnalyzer.java, checked if the JOIN expression needed a warning and created one if necessary
* In TestAnalyzer, added JUnit tests that check for warning when OR is directly in predicate and no warning when OR not present

### CODE REVIEW

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/LogicalBinaryExpression.java [hunk 1]
```diff
@@ -82,6 +82,24 @@ public Expression getRight()
         return right;
     }
 
+    public boolean needsJoinWarning()
```
Comment (tdcmeehan) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/LogicalBinaryExpression.java:85]: The operator has public visibility, so this doesn't need to be a method on this class.  I'd put it as a helper method in `StatementAnalyzer`.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/LogicalBinaryExpression.java [hunk 2]
```diff
@@ -82,6 +82,24 @@ public Expression getRight()
...
+    {
+        if (this.operator == Operator.OR) {
```
Comment (tdcmeehan) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/LogicalBinaryExpression.java:87]: Consider implementing this with a `ArrayDequeue<Expression>` to remove the recursion.

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java [hunk 3]
```diff
@@ -92,6 +92,12 @@ private static void assertHasWarning(WarningCollector warningCollector, Standard
         assertTrue(warning.getMessage().startsWith(match));
     }
 
+    private static void assertNoWarning(WarningCollector warningCollector)
+    {
+        List<PrestoWarning> warnings = warningCollector.getWarnings();
+        assertEquals(warnings.size(), 0);
```
Comment (tdcmeehan) [inline presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java:98]:
```suggestion
        assertTrue(warnings.isEmpty());
```

**Review by tdcmeehan — COMMENTED** [review]
Just some nits.

**Comment by tdcmeehan** [comment]
This is the convention we follow for Git commit messages: https://chris.beams.io/posts/git-commit/

tl;dr, let's reword the Git commit message to be in the imperitive mood, and keep it concise to the feature being added.  A suggestion is `Add a warning for join with an or operator in the predicate`.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -1350,6 +1358,31 @@ else if (criteria instanceof JoinOn) {
             return output;
         }
 
+        private boolean needsJoinWarning(LogicalBinaryExpression expression)
+        {
+            ArrayDeque<Expression> queue = new ArrayDeque<Expression>();
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1363]: Nit: make the lefthand side just a `Deque<Expression>`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 5]
```diff
@@ -1350,6 +1358,31 @@ else if (criteria instanceof JoinOn) {
             return output;
         }
 
+        private boolean needsJoinWarning(LogicalBinaryExpression expression)
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1361]: Consider naming this `expressionContainsOr`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 6]
```diff
@@ -1350,6 +1358,31 @@ else if (criteria instanceof JoinOn) {
...
+                        return true;
+                    }
+
+                    queue.add(logicalExpression.getLeft());
+                    queue.add(logicalExpression.getRight());
+                }
+            }
+
+            return false;
+        }
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1384]: I would simplify a bit by just making the parameter an `Expression`, and letting the BFS handle whether or not to traverse if the root is not a `LogicalBinaryExpression`.  And then in the above code that uses it we can remove the `instanceof` check and cast, and just use the inner if check.

```suggestion
        private boolean expressionContainsOr(Expression expression)
        {
            Deque<Expression> queue = new ArrayDeque<Expression>();
            queue.offer(expression);

            while (!queue.isEmpty()) {
                Expression currentExpression = queue.poll();
                if (currentExpression instanceof LogicalBinaryExpression) {
                    LogicalBinaryExpression logicalExpression = (LogicalBinaryExpression) currentExpression;

                    if (logicalExpression.getOperator() == LogicalBinaryExpression.Operator.OR) {
                        return true;
                    }

                    queue.offer(logicalExpression.getLeft());
                    queue.offer(logicalExpression.getRight());
                }
            }

            return false;
        }
```

**Review by tdcmeehan — APPROVED** [review]
Just some nits, otherwise LGTM.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 4 above)
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1363]: queue -> expressionQueue

Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1363]: why do we need Deque here, a normal Queue is sufficient.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 7]
```diff
@@ -1350,6 +1358,31 @@ else if (criteria instanceof JoinOn) {
             return output;
         }
 
+        private boolean needsJoinWarning(LogicalBinaryExpression expression)
+        {
+            ArrayDeque<Expression> queue = new ArrayDeque<Expression>();
+            Expression currentExpression;
+            LogicalBinaryExpression logicalExpression;
+
+            queue.add(expression);
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1367]:
```suggestion
            queue.offer(expression);
```

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 8]
```diff
@@ -1350,6 +1358,31 @@ else if (criteria instanceof JoinOn) {
...
+
+            while (!queue.isEmpty()) {
+                currentExpression = queue.remove();
```
Comment (swapsmagic) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1370]:
```suggestion
                currentExpression = queue.poll();
```

**Review by swapsmagic — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 4 above)
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1363]: Good point, I forgot `Dequeue` implements `Queue`, and we just need queue semantics.

Comment (swapsmagic) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1363]: Just make it Queue<Expression> is sufficient.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 9]
```diff
@@ -1337,6 +1339,12 @@ else if (criteria instanceof JoinOn) {
                     // coerce null to boolean
                     analysis.addCoercion(expression, BOOLEAN, false);
                 }
+                if (expression instanceof LogicalBinaryExpression) {
+                    LogicalBinaryExpression logicalExpression = (LogicalBinaryExpression) expression;
+                    if (needsJoinWarning(logicalExpression)) {
+                        warningCollector.add(new PrestoWarning(PERFORMANCE_WARNING, "JOIN conditions with an OR may lead to a cross join with filter"));
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1345]: 1. Make it more explicit by saying - 'can cause performance issues as it may lead to a cross join with filter'
2. See if you can include the line/column number for the OR in the message

**Review by kaikalur — CHANGES_REQUESTED** [review]

**Comment by tdcmeehan** [comment]
Also, the comments in the commit message don't follow the commit guidelines.  I would just remove them, the title is sufficient for context.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 9 above)
Comment (prithvip) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: +1 on both. And maybe we can add something prescriptive so the user can fix their query. Something like "Consider rewriting this with a UNION..."

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 10]
```diff
@@ -1350,6 +1362,29 @@ else if (criteria instanceof JoinOn) {
             return output;
         }
 
+        private Expression findORExpression(Expression expression)
```
Comment (prithvip) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1365]: It's a bit awkward having a BFS here. For a cleaner approach, take a look at ExpressionTreeUtils, I am pretty sure you can replace this whole BFS with a one-liner.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 4 above)
Comment (chi-tsai) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Fixed

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 6 above)
Comment (chi-tsai) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Fixed

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 4 above)
Comment (chi-tsai) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Fixed

Comment (chi-tsai) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Fixed

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 7 above)
Comment (chi-tsai) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Fixed

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 8 above)
Comment (chi-tsai) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Fixed

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 9 above)
Comment (chi-tsai) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Done for both!

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 11]
```diff
@@ -1338,6 +1342,14 @@ else if (criteria instanceof JoinOn) {
                     analysis.addCoercion(expression, BOOLEAN, false);
                 }
 
+                Expression orExpression = findORExpression(expression);
+
+                if (orExpression != null) {
+                    NodeLocation orLocation = orExpression.getLocation().get();
+                    String warningMessage = format("line %s:%s: %s", orLocation.getLineNumber(), orLocation.getColumnNumber(), "JOIN conditions with an OR can cause performance issues as it may lead to a cross join with filter");
```
Comment (prithvip) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1349]: Moving this string format logic to a helper would encourage re-use. There are other places in this analyzer where it could be very nice to include line/column numbers in performance warnings.

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java [hunk 12]
```diff
@@ -110,6 +116,26 @@ public void testNonComparableWindowOrder()
         assertFails(TYPE_MISMATCH, "SELECT row_number() OVER (ORDER BY t.x) FROM (VALUES(color('red'))) AS t(x)");
     }
 
+    @Test
+    public void testORWarning()
```
Comment (prithvip) [inline presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java:120]: What about IN conditions? These two statements are equivalent, but we would throw a warning for one and not the other:
`SELECT * FROM t1 JOIN t2 ON t1.a = t2.a OR t1.a = t2.b`
`SELECT * FROM t1 JOIN t2 ON t1.a IN (t2.a, t2.b)`

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java [hunk 13]
```diff
@@ -110,6 +116,26 @@ public void testNonComparableWindowOrder()
...
+                PERFORMANCE_WARNING, "line 1:41: JOIN conditions with an OR can cause performance issues as it may lead to a cross join with filter");
+        assertHasWarning(analyzeWithWarnings("SELECT * FROM t1 JOIN t2 ON t1.a = t2.a \n" + "AND (t1.b = t2.b OR t1.b > t2.b)"),
+                PERFORMANCE_WARNING, "line 2:18: JOIN conditions with an OR can cause performance issues as it may lead to a cross join with filter");
+    }
+
+    @Test void testNoORWarning()
+    {
+        assertNoWarning(analyzeWithWarnings("SELECT * FROM t1 JOIN t2 ON t1.a = t2.a"));
+        assertNoWarning(analyzeWithWarnings("SELECT * FROM t1 JOIN t2 ON t1.a = t2.a AND t1.b = t2.b"));
+        assertNoWarning(analyzeWithWarnings("SELECT * FROM t1 JOIN t2 ON t1.a = t2.a AND IF(t2.b = t1.a OR t2.b = null, 'YES', 'NO') = 'YES'"));
```
Comment (prithvip) [inline presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java:136]: Why should this not throw a warning? These are equivalent:
`IF(t2.b = t1.a OR t2.b = null, 'YES', 'NO') = 'YES'`
`t2.b = t1.a OR t2.b = null`

**Review by prithvip — CHANGES_REQUESTED** [review]

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java (hunk 13 above)
Comment (kaikalur) [reply presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java:136]: @prithvip @prithvip it's a bit hard to do this simplification. @chi-tsai  wrote a really good test case but unfortunately we can't detect it right now and anyway it won't be a cross join because the there is still one key condition (t1.a = t2.). So I think for now it's good to err on the side of giving fewer false positives like:

T1.x = T2.x AND IF(T1.y = 'a' OR T1.z = 2, w, null) = T2.w 

I think that's ok for now.

Maybe later on, we can try and move this check to the planner where we do DNF and CNF so we can better error messages.

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java (hunk 12 above)
Comment (kaikalur) [reply presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java:120]: We can add this as a separate rule. So for now it's fine.

**Comment by kaikalur** [comment]
@prithvip for this rule we really want to catch the most common and egregious use cases so we can ask the DIET team to surface them so not risking too many things like:

T1.x = T2.x AND IF(T1.y = 'a' OR T1.z = 2, w, null) = T2.w 

which we don't want to warn

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 11 above)
Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1349]: Yeah - we can open a separate issue for refactoring this analyzer warning/error handling.

**Review by kaikalur — CHANGES_REQUESTED** [review]

**Review by swapsmagic — APPROVED** [review]
LGTM. Let's get Travis build green and it should be good.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 10 above)
Comment (prithvip) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1371]: If this query were:
`SELECT * FROM t1 JOIN t2 ON t1.a = t2.a AND (t2.b = t1.a OR t2.b = null)`
This would throw a warning. But should it? 
Since there is a key condition t1.a=t2.a, this shouldn't be a cross join. 
So I think the existing implementation with the BFS has the same issue.

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java (hunk 13 above)
Comment (prithvip) [reply presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java:137]: @kaikalur I think that this way of warning will be confusing for users, because it is unclear what is actually being warned against. One of the test cases for the warning is this query:
`SELECT * FROM t1 JOIN t2 ON t1.a = t2.a AND (t1.b = t2.b OR t1.b > t2.b)`
There is a key condition here too, just like in your example. 
I agree that it is more difficult to do more sophisticated detection of cross joins. 
So in the absence of that, we should stick with a simpler definition for the warning, that triggers if there is any OR condition present in the join condition predicates, even if it is in a conditional or other construct.

Comment (rongrong) [reply presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java:137]: If we want to warn people about cross join, why don't we check this in JoinNode instead after planning? You'd potentially have some expressions pushed down to simplify it (I didn't think this through though).

**Comment by prithvip** [comment]
I think we need clarity on the definition of this warning. Right now, the definition is something like "if the join condition is composed only of ANDs/ORs, and there is an OR, then throw". But I think this is confusing because, even though the warning message is about a cross join, it would trigger for queries that are not cross joins, for example `SELECT * FROM t1 JOIN t2 ON t1.a = t2.a AND (t1.b = t2.b OR t1.b > t2.b)`. 

One way to make things clearer is to define the warning as "join condition contains an OR somewhere". But, even if this is simpler warning to explain, this might lead to more false positives, not less. 

A better approach might be to narrow the warning to only throw if the outermost join condition is an OR. It is still possible such a query might not end up being a cross join, since there still could be a key equality after the expression is transformed. But I think it would only be a limited number of false positives, and we can tweak the warning message.

**Comment by kaikalur** [comment]
> I think we need clarity on the definition of this warning. Right now, the definition is something like "if the join condition is composed only of ANDs/ORs, and there is an OR, then throw". But I think this is confusing because, even though the warning message is about a cross join, it would trigger for queries that are not cross joins, for example `SELECT * FROM t1 JOIN t2 ON t1.a = t2.a AND (t1.b = t2.b OR t1.b > t2.b)`.
> 
> One way to make things clearer is to define the warning as "join condition contains an OR somewhere". But, even if this is simpler warning to explain, this might lead to more false positives, not less.
> 
> A better approach might be to narrow the warning to only throw if the outermost join condition is an OR. It is still possible such a query might not end up being a cross join, since there still could be a key equality after the expression is transformed. But I think it would only be a limited number of false positives, and we can tweak the warning message.

Thanks for the great explanation and thorough investigation, @prithvip! I agree that makes a lot sense. 

@chi-tsai  just update the PR to do what Prithvi suggested and then we can review it again.

**Review by tdcmeehan — APPROVED** [review]
LGTM

**Review by prithvip — APPROVED** [review]
LGTM. Thanks for doing this!
